src/Grpc.Net.Client/Balancer/Internal/SocketConnectivitySubchannelTransport.cs

```diff
@@ -199,75 +204,121 @@ public async ValueTask<ConnectResult> TryConnectAsync(ConnectContext context)
         {
             if (!_disposed)
             {
-                _socketConnectedTimer.Change(TimeSpan.Zero, TimeSpan.Zero);
+                _socketConnectedTimer.Change(Timeout.InfiniteTimeSpan, Timeout.InfiniteTimeSpan);
             }
         }
         return result;
     }
 
-    private async void OnCheckSocketConnection(object? state)
+    private void OnCheckSocketConnection(object? state)
     {
         try
         {
-            var socket = _initialSocket;
-            if (socket != null)
+            Socket? socket;
+            BalancerAddress? socketAddress;
+            var closeSocket = false;
+            Exception? checkException = null;
+
+            lock (Lock)
             {
-                CompatibilityHelpers.Assert(_initialSocketAddress != null);
+                socket = _initialSocket;
+                socketAddress = _initialSocketAddress;
 
-                var closeSocket = false;
-                Exception? sendException = null;
-                try
+                if (socket != null)
                 {
-                    // Check the socket is still valid by doing a zero byte send.
-                    SocketConnectivitySubchannelTransportLog.CheckingSocket(_logger, _subchannel.Id, _initialSocketAddress);
-                    await socket.SendAsync(Array.Empty<byte>(), SocketFlags.None).ConfigureAwait(false);
+                    CompatibilityHelpers.Assert(socketAddress != null);
 
-                    // Also poll socket to check if it can be read from.
-                    closeSocket = IsSocketInBadState(socket);
-                }
-                catch (Exception ex)
-                {
-                    closeSocket = true;
-                    sendException = ex;
-                    SocketConnectivitySubchannelTransportLog.ErrorCheckingSocket(_logger, _subchannel.Id, _initialSocketAddress, ex);
+                    try
+                    {
+                        SocketConnectivitySubchannelTransportLog.CheckingSocket(_logger, _subchannel.Id, socketAddress);
+
+                        // Poll socket to check if it can be read from.
+                        // We don't care that there is available data to be be read at this point. If there is available data
+                        // then the socket can't be used by SocketsHttpHandler, but the socket can get recreated when it is eventually used.
+                        //
+                        // Available data needs to be read now because the only way to determine whether the connection is closed is to
+                        // get the results of polling after available data is received.
+                        //
+                        // If a socket with data is closed immediately then that might happen continuously.
+                        // Chill, leave the socket alone, and be happy that it can be read from. We can create a new one later.
+                        do
```

Comment:
just to check my understanding; it looks like this only happens during initial connect, and we're intentionally burning the data (and triggering a failure condition) because this is unexpected, but we want to be able to poll later, and thus want an empty state - is that right?

quick question: why do we explicitly not expect data initially? is that a protocol specification thing, that the server always talks first?

Reply:
This issue arose because I didn't completely understand the HTTP/2 handshake process.

I thought a client always initializes an HTTP/2 connection by sending the HTTP/2 preface, and then the server responds with a SETTINGS frame. That's the way Kestrel/IIS/HTTP.sys work.

It turns out that some HTTP/2 server implementations don't wait for the preface. They preemptively send the SETTINGS frame. There isn't a rule against doing this.

See https://datatracker.ietf.org/doc/html/rfc7540#section-3.5